include/cache.h

```diff
@@ -13,45 +13,75 @@
 extern "C" {
 #endif
 
-#define _sys_cache_flush_sig(x) void (x)(vaddr_t virt, size_t size)
-
-#if defined(CONFIG_CACHE_FLUSHING)
-
-#if defined(CONFIG_ARCH_CACHE_FLUSH_DETECT)
-	typedef _sys_cache_flush_sig(_sys_cache_flush_t);
-	extern _sys_cache_flush_t *sys_cache_flush;
+/**
+ *
+ * @brief Flush d-cache lines to main memory
+ *
+ * No alignment is required for either addr or size, but since
+ * sys_cache_flush() iterates on the d-cache lines, a d-cache line alignment for
+ * both is optimal.
+ *
+ * The d-cache line size is specified either via the CONFIG_CACHE_LINE_SIZE
+ * kconfig option or it is detected at runtime.
+ *
+ * @param addr the pointer to start the multi-line flush
+ * @param size the number of bytes that are to be flushed
+ * @param flags architecture specific flag
+ *
+ * @return N/A
+ */
+static inline void sys_cache_flush(void *addr, size_t size, u32_t flags)
+{
+#ifdef CONFIG_CACHE_FLUSHING
+	arch_cache_flush(addr, size, flags);
 #else
-	extern _sys_cache_flush_sig(sys_cache_flush);
+	ARG_UNUSED(addr);
+	ARG_UNUSED(size);
+	ARG_UNUSED(flags);
 #endif
+}
 
-#else
-
-/*
- * Provide NOP APIs for systems that do not have caches.
+/**
+ *
+ * @brief Invalidate d-cache lines
```

Comment:
I think some supporting comments should be here, like has been done with `sys_cache_flush()`